CPAC/pipeline/engine/resource_pool.py

```diff
@@ -1412,1351 +1416,54 @@ def node_data(self, resource, **kwargs):
-            map_node=True
-        )
-        cfg.set_nested(cfg, key, node)
-    """
-
-    return rpool
-
-
-def initiate_rpool(wf, cfg, data_paths=None, part_id=None):
-    """
-    Initialize a new ResourcePool.
-
-    data_paths format:
-      {'anat': {
-            'T1w': '{T1w path}',
-            'T2w': '{T2w path}'
-        },
-       'creds_path': {None OR path to credentials CSV},
-       'func': {
-           '{scan ID}':
-               {
-                   'scan': '{path to BOLD}',
-                   'scan_parameters': {scan parameter dictionary}
-               }
-       },
-       'site_id': 'site-ID',
-       'subject_id': 'sub-01',
-       'unique_id': 'ses-1',
-       'derivatives_dir': '{derivatives_dir path}'}
-    """
-    # TODO: refactor further, integrate with the ingress_data functionality
-    # TODO: used for BIDS-Derivatives (below), and possible refactoring of
-    # TODO: the raw data config to use 'T1w' label instead of 'anat' etc.
-
-    if data_paths:
-        part_id = data_paths["subject_id"]
-        ses_id = data_paths["unique_id"]
-        if "creds_path" not in data_paths:
-            creds_path = None
-        else:
-            creds_path = data_paths["creds_path"]
-        unique_id = f"{part_id}_{ses_id}"
-
-    elif part_id:
-        unique_id = part_id
-        creds_path = None
-
-    rpool = ResourcePool(name=unique_id, cfg=cfg)
-
-    if data_paths:
-        # ingress outdir
-        try:
-            if (
-                data_paths["derivatives_dir"]
-                and cfg.pipeline_setup["outdir_ingress"]["run"]
-            ):
-                wf, rpool = ingress_output_dir(
-                    wf,
-                    cfg,
-                    rpool,
-                    unique_id,
-                    data_paths,
-                    part_id,
-                    ses_id,
-                    creds_path=None,
-                )
-        except:
-            rpool = ingress_raw_anat_data(
-                wf, rpool, cfg, data_paths, unique_id, part_id, ses_id
-            )
-            if "func" in data_paths:
-                wf, rpool, diff, blip, fmap_rp_list = ingress_raw_func_data(
-                    wf, rpool, cfg, data_paths, unique_id, part_id, ses_id
-                )
-
-    # grab any file paths from the pipeline config YAML
-    rpool = ingress_pipeconfig_paths(cfg, rpool, unique_id, creds_path)
-
-    # output files with 4 different scans
-
-    return (wf, rpool)
-
-
-def run_node_blocks(blocks, data_paths, cfg=None):
-    import os
-
-    from CPAC.pipeline import nipype_pipeline_engine as pe
-    from CPAC.pipeline.engine import NodeBlock
+                CpacProvenance = default_CpacProvenance
+            resource = Resource(row, CpacProvenance)
+            # making the rpool a list so that the duplicates are appended rather than overwritten
+            self.rpool.setdefault(resource.suffix, [])
+            self.rpool[resource.suffix].append(resource)
```

Comment:
I know this branch isn't done yet, but I don't understand the design of this change. We want duplicates? We're changing the API how we access resources from the pool?